File: RMQ/sqrtDescompositionDynamic.cpp
#include <bits/stdc++.h>
using namespace std;

const int INF = 1e9 + 1;

int main() {
    int n, q;
    cin >> n >> q;

    int a[n];
    for (int i = 0; i < n; i++)
        scanf("%d", &a[i]);

    int s = sqrt(n) + 1;
    int b[s];
    for (int i = 0; i < s; i++)
        b[i] = INF;

    for (int i = 0; i < n; i++)
        b[i / s] = min(b[i / s], a[i]);

    while (q--) {
        int type;
        scanf("%d", &type);

        if (type == 1) {
            int k, u;
            scanf("%d %d", &k, &u);

            k--;

            a[k] = u;

            int block_index = k / s;
            b[block_index] = INF;

            for (int i = block_index*s; i < min(n, (block_index+1)*s); i++)
                b[block_index] = min(b[block_index], a[i]);
        }
        else {
            int l, r;
            scanf("%d %d", &l, &r);
            
            l--, r--;

            int ans = INF;
            int i = l;

            while (i <= r) {
                if (i % s == 0 && i + s - 1 <= r) {
                    ans = min(ans, b[i / s]);
                    i += s;
                }
                else {
                    ans = min(ans, a[i]);
                    i++;
                }
            }

            printf("%d\n", ans);
        }
    }
}

File: sqrtDescompositionDynamic.py
import sys
input = sys.stdin.readline

INF = 10**9 + 1
block_size = 400

n, q = map(int, input().split())
a = list(map(int, input().split()))

block_number = (n + block_size - 1) // block_size

b = [INF] * block_number

for i in range(n):
    block_index = i // block_size
    b[block_index] = min(b[block_index], a[i])

answers = []

for i in range(q):
    type, x, y = map(int, input().split())

    if type == 1:
        k = x - 1
        u = y

        a[k] = u
        block_index = k // block_size 
        
        b[block_index] = INF
        for j in range(block_index*block_size, min((block_index+1) * block_size , n)):
            b[block_index] = min(b[block_index], a[j])
    else:
        l = x - 1
        r = y - 1

        j = l
        ans = INF
    
        while j <= r:
            if j % block_size == 0 and j + block_size - 1 <= r:
                ans = min(ans, b[j // block_size])
                j += block_size
            else:
                ans = min(ans, a[j])
                j += 1
        
        answers.append(str(ans))

# Output all answers
sys.stdout.write("\n".join(answers))
sys.stdout.write("\n")